File: text-span-bullet/src/main/java/net/xpece/android/text/span/BulletSpanCompatFactory.java
package net.xpece.android.text.span;

import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;
import androidx.annotation.Px;

import kotlin.jvm.functions.Function4;

import static android.os.Build.VERSION.SDK_INT;
import static android.text.style.BulletSpan.STANDARD_GAP_WIDTH;

@SuppressWarnings("deprecation")
final class BulletSpanCompatFactory {

    // Bullet is slightly bigger to avoid aliasing artifacts on mdpi devices.
    private static final int STANDARD_BULLET_RADIUS = 4;

    private static final Function4<Integer, Integer, Boolean, Integer, BulletSpanCompat> INSTANCE;

    static {
        if (SDK_INT >= 28) {
            INSTANCE = BulletSpanPie::create;
        } else {
            INSTANCE = BulletSpanKitkat::new;
        }
    }

    static BulletSpanCompat create(
            @Px int gapWidth,
            @Nullable @ColorInt Integer color,
            @Px int bulletRadius) {
        boolean wantColor = color != null;
        return INSTANCE.invoke(gapWidth, wantColor ? color : 0, wantColor, bulletRadius);
    }

    static BulletSpanCompat create(
            @Px int gapWidth,
            @ColorInt int color) {
        return create(gapWidth, color, STANDARD_BULLET_RADIUS);
    }

    static BulletSpanCompat create(
            @Px int gapWidth) {
        return create(gapWidth, null, STANDARD_BULLET_RADIUS);
    }

    static BulletSpanCompat create() {
        return create(STANDARD_GAP_WIDTH, null, STANDARD_BULLET_RADIUS);
    }

    private BulletSpanCompatFactory() {
        throw new AssertionError();
    }
}

File: text-span/src/main/java/net/xpece/android/text/span/BulletSpanCompatFactory.kt
@file:JvmName("BulletSpanCompatFactory")
@file:Suppress("FunctionName")
@file:RestrictTo(RestrictTo.Scope.LIBRARY)

package net.xpece.android.text.span

import android.os.Build.VERSION.SDK_INT
import android.os.Parcel
import androidx.annotation.ColorInt
import androidx.annotation.Px
import androidx.annotation.RequiresApi
import androidx.annotation.RestrictTo

@RestrictTo(RestrictTo.Scope.LIBRARY)
internal fun interface BulletSpanFactory {

    operator fun invoke(
        @Px gapWidth: Int,
        @ColorInt color: Int,
        wantColor: Boolean,
        @Px bulletRadius: Int,
    ): BulletSpanCompat

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    companion object {

        @get:JvmName("getInstance")
        val BulletSpanCompatFactory: BulletSpanFactory = when {
            SDK_INT >= 28 -> BulletSpanPieFactory
            else -> BulletSpanKitkatFactory
        }
    }
}

private object BulletSpanKitkatFactory : BulletSpanFactory {

    override fun invoke(
        gapWidth: Int,
        color: Int,
        wantColor: Boolean,
        bulletRadius: Int
    ): BulletSpanCompat {
        return BulletSpanKitkat(gapWidth, color, wantColor, bulletRadius)
    }
}

@RequiresApi(28)
private object BulletSpanPieFactory : BulletSpanFactory {

    override fun invoke(
        gapWidth: Int,
        color: Int,
        wantColor: Boolean,
        bulletRadius: Int
    ): BulletSpanCompat {
        return try {
            BulletSpanParcel(gapWidth, color, wantColor, bulletRadius)
        } catch (_: Throwable) {
            BulletSpanPie(gapWidth, color, wantColor, bulletRadius)
        }
    }

    private fun BulletSpanParcel(
        gapWidth: Int,
        color: Int,
        wantColor: Boolean,
        bulletRadius: Int,
    ): BulletSpanCompat {
        val parcel = Parcel.obtain()
        try {
            parcel.writeInt(gapWidth)
            parcel.writeInt(color)
            parcel.writeInt(if (wantColor) 1 else 0)
            parcel.writeInt(bulletRadius)
            parcel.setDataPosition(0)
            return BulletSpanPie(parcel)
        } finally {
            parcel.recycle();
        }
    }
}
